### ISSUE TITLE
Datatype in manually boolean-marked variables not propagated

### ISSUE THREAD

**Issue by RootCubed** [NationalSecurityAgency/ghidra#4457]
**Describe the bug**
If the decompiler is unable to infer the boolean type of a variable on its own and the user manually marks it as `bool`, the decompiler is unable to propagate this information further and is therefore unable to apply any boolean-specific simplification rules such as `RuleBooleanNegate`.

(If the compiler is able to infer the boolean type on its own, the boolean-specific simplification rules are applied as expected)

**To Reproduce**
1. Compile and import any of the example programs provided below.
2. Observe that the output that Ghidra generates does not match the expected output.

**Attachments**
Example program 1:
```c
int test1(bool param_1) {
    if (param_1) return 42;
    return 43;
}
```

Here, Ghidra emits the if statement as `if (param_1 == false)`, whereas boolean rules should be able to simplify this to `if (param_1)`.

Example program 2:
```c
bool test2(int a) {
    bool res = false;
    if (a == 5) {
        res = test2(a + 1);
    }
    return !res;
}
```

In this example, Ghidra emits the final statement as `return (bool)(res ^ 1);`, whereas boolean rules should be able to simplify this to `return !res`.

**Environment**
- OS: Windows 11
- Java Version: 15.0.1
- Ghidra Version: 10.1.4
- Ghidra Origin: official GitHub build

**Additional context**
Original issue: #4338

**Comment by ryanmkurtz** [comment]
Thanks!

**Comment by Wall-AF** [comment]
Another good example is the Windows API type `BOOL` which is actually 2-bytes for me.

**Comment by astrelsky** [comment]
> Another good example is the Windows API type `BOOL` which is actually 2-bytes for me.

It's a typedef to an unsigned int. It is not a boolean and these rules cannot be applied.

**Comment by Wall-AF** [comment]
> > Another good example is the Windows API type `BOOL` which is actually 2-bytes for me.
> 
> It's a typedef to an unsigned int. It is not a boolean and these rules cannot be applied.

Yes, but, in my case, I've replaced that with an `enum` with `TRUE` & `FALSE` values!

**Comment by RootCubed** [comment]
@caheckman Unfortunately, it seems that example program 2 from the original post is still not producing the expected output.
Additionally, I came across a much more common case where this issue also appears, namely if a structure contains a boolean.
This example should demonstrate it pretty well:
```cpp
struct S {
    bool bVal;
};

bool func(S *s) {
    return !s->bVal;
}
```

Currently, the decompiler outputs the following:
```cpp
bool func(S *s){
  return (bool)(s->bVal ^ 1);
}
```

**Comment by RootCubed** [comment]
Replacing the body of `Varnode::isBooleanValue` with simply `return (size == 1 && type->getMetatype() == TYPE_BOOL);` "fixes" the example in my previous comment. That will almost certainly cause regressions somewhere else, but I thought I'd mention this anyway; hopefully it can help with figuring out what needs to be changed in order to fix this issue.

Also, since it didn't fix the second example in my original comment, this might actually be a slightly separate, but related issue. If you'd prefer it if I made a separate issue for this case with the structure, just let me know.

**Comment by RootCubed** [comment]
@caheckman I have tried out a few different approaches to fix this now, but in the end it always comes back to trying to understand what exactly `PcodeOp::isCalculatedBool()` is supposed to denote. It seems to me that if the `calculated_bool` flag is set, then the metatype of the output Varnode is always boolean too anyways.

So my question is, in what situations would we not want to trust that type annotation / do we have the `calculated_bool` flag not set even though the type annotation says it's a bool?

I'd be very grateful if you could clarify this for me :)

So one possible fix that has as few behavior changes as possible would be to change the definition of `Varnode::isBooleanValue` to:
```cpp
bool Varnode::isBooleanValue(bool useAnnotation) const

{
  if (isWritten()) {
    if (useAnnotation && def->code() == CPUI_LOAD && size == 1 && type->getMetatype() == TYPE_BOOL)
      return true;
    return def->isCalculatedBool();
  }
  if (!useAnnotation)
    return false;
  if ((flags & (input | typelock)) == (input | typelock)) {
    if (size == 1 && type->getMetatype() == TYPE_BOOL)
      return true;
  }
  return false;
}
```

I also tried to do `opMarkCalculatedBool` inside of `TypeOpLoad::propagateType`, but we don't have access to `Funcdata` inside of there...
